Allow variables to be used when initing neighbor stateful props

lindsayad opened the pull request:
Previously we did not reinit the variables so you would get a bad access
if you used a variable when initializing the neighbor version of
stateful material properties. I'm honestly a little surprised we never
hit this before

Closes #19735

lindsayad committed:
Allow variables to be used when initing neighbor stateful props

Previously we did not reinit the variables so you would get a bad access
if you used a variable when initializing the neighbor version of
stateful material properties. I'm honestly a little surprised we never
hit this before

Closes #19735

lindsayad committed:
test

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/922214/) on e233a99 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19918/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/922224/) on e233a99 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>626704</th>
      <th colspan=4>#19918 e233a9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.18%</td>
      <td>82.19%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72924</td>
      <td>72925</td>
      <td>+1</td>
      <td>9</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15808</td>
      <td>15807</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19918/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19918/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19918/coverage/xfem)

</details>



This comment will be updated on new commits.

fdkong commented on framework/src/loops/ComputeMaterialsObjectThread.C:
```diff
@@ -171,7 +171,15 @@ ComputeMaterialsObjectThread::onInternalSide(const Elem * elem, unsigned int sid
   {
-    _assembly[_tid]->reinit(elem, side);
+    const Elem * neighbor = elem->neighbor_ptr(side);
+    const dof_id_type elem_id = elem->id(), neighbor_id = neighbor->id();
```
Do we need to check if neighbor is empty?  Or this has been guarded by `_need_internal_side_material` ?

lindsayad replied:
In `ThreadedElementLoopBase` we do not call `onInternalSide` if the neighbor is null

fdkong commented on framework/src/loops/ComputeMaterialsObjectThread.C:
```diff
@@ -171,7 +171,15 @@ ComputeMaterialsObjectThread::onInternalSide(const Elem * elem, unsigned int sid
-    _assembly[_tid]->reinit(elem, side);
+    const Elem * neighbor = elem->neighbor_ptr(side);
+    const dof_id_type elem_id = elem->id(), neighbor_id = neighbor->id();
+    const bool should_continue =
```
It is a bit hard to understand this logic. Could you add comments ?

loganharbour replied:
Can we just add this logic as a utility somewhere if we use it multiple times? I think we use the same logic in DG? Maybe something on MooseMesh?

fdkong commented on test/src/materials/VarCouplingMaterial.C:
```diff
@@ -21,12 +21,16 @@ VarCouplingMaterial::validParams()
+    _var(getParam<bool>("use_tag") ? coupledVectorTagValue("var", "tag") : coupledValue("var")),
     _base(getParam<Real>("base")),
     _coef(getParam<Real>("coef")),
     _diffusion(declareProperty<Real>("diffusion")),
```
This is not related to this PR. I was curious about why we call the material "diffusion"?

This should be a general material

lindsayad replied:
I'm guessing it was probably because the test creator was coming up with a test where the material property was being consumed by a diffusion object

fdkong reviewed (changes requested):
Minor comments

lindsayad committed:
Address Fande review

loganharbour reviewed (dismissed):
Looks good to me otherwise. Up to you if you wanna cleanup duplication in skip logic (if there is duplication)

lindsayad committed:
Consolidate should-compute logic into ThreadedElementLoopBase

The majority of our loops use the same logic to determine whether to
evaluate an internal face so I think it makes sense to have that logic
in the base class and to override when necessary in derived loops

loganharbour reviewed (approved):
Beautiful code 👍🏼

lindsayad commented:
@fdkong does this look fine for you? Please resolve conversations if they're resolved for you

fdkong reviewed (approved):
Nice improvement
